In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime

In [16]:
# if the file is not read, try to save it as a csv utf-18
df = pd.read_csv("Online_Retail_data.csv")
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12-01-2010 08:26       2.55     17850.0  United Kingdom  
1       12-01-2010 08:26       3.39     17850.0  United Kingdom  
2       12-01-2010 08:26       2.75     17850.0  United Kingdom  
3       12-01-2010 08:26       3.39     17850.0  United Kingdom  
4       12-01-2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12-09-2011 12:50       0.85     12680.0          France  
541905  12-09-2011 12:50       2.10     12680.0          France  
541906  12-09-2011 12:50       4.15     12680.0          France  
541907  12-09-2011 12:50       4.15     12680.0          France  
541908  12-09-2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

---

### Get the count of null values

In [17]:
df.notna().sum()

InvoiceNo      541909
StockCode      541909
Description    540455
Quantity       541909
InvoiceDate    541909
UnitPrice      541909
CustomerID     406829
Country        541909
dtype: int64

---

### Removing rows with null values

In [18]:
new_df = df.dropna()
new_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12-01-2010 08:26       2.55     17850.0  United Kingdom  
1       12-01-2010 08:26       3.39     17850.0  United Kingdom  
2       12-01-2010 08:26       2.75     17850.0  United Kingdom  
3       12-01-2010 08:26       3.39     17850.0  United Kingdom  
4       12-01-2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12-09-2011 12:50       0.85     12680.0          France  
541905  12-09-2011 12:50       2.10     12680.0          France  
541906  12-09-2011 12:50       4.15     12680.0          France  
541907  12-09-2011 12:50       4.15     12680.0          France  
541908  12-09-2011 12:50       4.95     12680.0          France  

[406829 rows x 8 columns]

---

### Check if there is null value

In [19]:
new_df.isna().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

---

### Get the count of duplicate rows

In [20]:
new_df.duplicated().sum()

np.int64(5225)

---

### Drop the duplicate rows

In [21]:
new_df = new_df.drop_duplicates()
new_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12-01-2010 08:26       2.55     17850.0  United Kingdom  
1       12-01-2010 08:26       3.39     17850.0  United Kingdom  
2       12-01-2010 08:26       2.75     17850.0  United Kingdom  
3       12-01-2010 08:26       3.39     17850.0  United Kingdom  
4       12-01-2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12-09-2011 12:50       0.85     12680.0          France  
541905  12-09-2011 12:50       2.10     12680.0          France  
541906  12-09-2011 12:50       4.15     12680.0          France  
541907  12-09-2011 12:50       4.15     12680.0          France  
541908  12-09-2011 12:50       4.95     12680.0          France  

[401604 rows x 8 columns]

### Count check for duplicate data
##### Earlier df had 541909 total rows 
##### The new_df had 406829 rows after removing null/NA
##### After removing duplicates they are 401604

In [22]:
new_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12-01-2010 08:26       2.55     17850.0  United Kingdom  
1       12-01-2010 08:26       3.39     17850.0  United Kingdom  
2       12-01-2010 08:26       2.75     17850.0  United Kingdom  
3       12-01-2010 08:26       3.39     17850.0  United Kingdom  
4       12-01-2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12-09-2011 12:50       0.85     12680.0          France  
541905  12-09-2011 12:50       2.10     12680.0          France  
541906  12-09-2011 12:50       4.15     12680.0          France  
541907  12-09-2011 12:50       4.15     12680.0          France  
541908  12-09-2011 12:50       4.95     12680.0          France  

[401604 rows x 8 columns]

---

### Updating datetime format

In [23]:
new_df["InvoiceDate"].dtype

dtype('O')

In [10]:
new_df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

In [24]:
# new_df["InvoiceDate"] = pd.to_datetime(new_df["InvoiceDate"], format="%d-%m-%Y %H:%M")
# this method should work for a consistent data format "12/6/10 9:26"
# But here the string varies like "12-6-10 9:26", so its, inconsistent and raises error

# let pandas detect each row individually, as each row has its own format  
pd.to_datetime(new_df["InvoiceDate"], format="mixed", dayfirst=False, errors="coerce")

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 401604, dtype: datetime64[ns]

In [25]:
new_df["InvoiceDate"] = pd.to_datetime(new_df["InvoiceDate"], format="mixed", dayfirst=False, errors="coerce")
new_df

C:\Users\Jigu\AppData\Local\Temp\ipykernel_12864\3624101072.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["InvoiceDate"] = pd.to_datetime(new_df["InvoiceDate"], format="mixed", dayfirst=False, errors="coerce")


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[401604 rows x 8 columns]

---

### Adding Total Amount Column

In [26]:
new_df["TotalAmount"] = new_df["Quantity"] * new_df["UnitPrice"]
new_df

C:\Users\Jigu\AppData\Local\Temp\ipykernel_12864\3516597197.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["TotalAmount"] = new_df["Quantity"] * new_df["UnitPrice"]


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  TotalAmount  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom        15.30  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom        22.00  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
...                    ...        ...         ...             ...          ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France        10.20  
541905 2011-12-09 12:50:00       2.10     12680.0          France        12.60  
541906 2011-12-09 12:50:00       4.15     12680.0          France        16.60  
541907 2011-12-09 12:50:00       4.15     12680.0          France        16.60  
541908 2011-12-09 12:50:00       4.95     12680.0          France        14.85  

[401604 rows x 9 columns]